### PR TITLE
Admin Panel for Configurable Navigational Links

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
_(Backend Only, Frontend PR in a subsequent PR)_

Currently we have a couple "key links" areas. (left sidebar on home page, site footer, key links page.

We should be able to define what these links are at the app level, both for generalization purposes and also possibly for dev.to administration simplification. Admins of Forems  should have the ability to add to the navigation sidebar links with additional community products such as books, movies etc., whilst being able to remove the links that are specific to DEV or those that they do not want.

The configurable elements of a navigation item is:
- Friendly name
- Icon (the podcast icon as an example).
- The destination page url
- A value that sets whether the link should be shown in the sidebar 'pre auth' or not.

You'll see that this PR was a journey and a **huge learning experience** for me:🛣
 At first I tried to fit the navigation links within the Site Config page and model, however as I started working with the more complicated actions like "update", I began to understand @citizen428's foresighted early-on comments (see below for comments). The navigation links are not simple scalar values or key-value pairs, but it is essentially an array of hashes which adds unnecessary complexity to work with within the Site Config. 
Hence it made more sense to pull the configurable Navigation Links into it's own page (in the future we'll probably move more sections out of the site config as its getting too complicated) and backed by its own table. In this way we could also add validations on the model without needing to go through a service.

This PR covers the admin portion of adding the database table + the admin page to configure the links. A subsequent PR will remove the hardcoded links and replace them with the values form the table. I chose to separate out the PR's so we can prep the database for DEV before we update the frontend :) 

## Related Tickets & Documents
https://github.com/forem/forem/issues/4661
https://github.com/forem/forem/issues/5475

## QA Instructions, Screenshots, Recordings
1. Go to http://localhost:3000/admin/navigation_links
2. Play around with the CRUD
3. Check that it saves correctly to the db 
 
**View:**
Screenshot 2020-10-01 at 16 07 26

**Create:**
Screenshot 2020-10-01 at 16 37 55

**Update:**
Screenshot 2020-10-01 at 16 37 48

**Delete:**
There will be a modal to verify deletion.

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?
Not right now

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/models/site_config.rb [hunk 1]
```diff
@@ -39,6 +41,11 @@ class SiteConfig < RailsSettings::Base
     members: ApplicationConfig["DEFAULT_EMAIL"]
   }
 
+  field :navigation, type: :array, default: [
+    { name: 'Reading List', url: 'abc.com', icon: LISTING_ICON },
+    { name: 'Listing', url: 'def.com', icon: MIC_ICON }
```
Comment (rhymes) [inline app/models/site_config.rb:46]: can the URLs be empty?

Comment (Ridhwana) [reply app/models/site_config.rb:46]: This is a good question, technically they shouldn't be empty but it would not break the interface if they are. We will have a set of default urls and navigation elements as a start.  

What are your thoughts?

Comment (rhymes) [reply app/models/site_config.rb:46]: maybe we can default them to the actual `/readinglist` and `/listings` ? 

ps. `Listing` should be plural: `Listings`

Comment (Ridhwana) [reply app/models/site_config.rb:46]: Haha yes definitely, this PR was just a spike to show that in theory hooking all this up works. I will definitely use sensible names, values and urls. The defaults will be as per the current dev website, not my random values  😅

**Comment by Ridhwana** [comment]
> **thought (non-blocking):** While I'm ok with this approach, I'm not entirely sure how I feel about putting more than scalar values or simple key-value pairs into `SiteConfig`.
> 
> We lose the ability to easily add validations and similar things, so maybe this could also be backed by an actual table (especially if we want to add more attributes) and we do the configuration through `/internal/navigation` (similar to `/internal/pages`).

@citizen428 I had to mull over this for a bit, I agree with your thought process that we lose the ability to easily add validations, and that's actually a pretty big deal. I'm not so sure though about doing the config through `/internal/navigation`, because then we have some config living on the internal/config and others on other routes. We'd need a way to bring it all back together again. I'm conflicted on whether to move forward with this PR and what approach to use.

**Comment by rhymes** [comment]
@Ridhwana what if we add https://api.rubyonrails.org/classes/ActiveModel/Validations.html ? would that work?

**Comment by Ridhwana** [comment]
> @Ridhwana what if we add https://api.rubyonrails.org/classes/ActiveModel/Validations.html ? would that work?

@rhymes I don't think it would work on this current PR implementation where we have an array of hashes.

**Comment by rhymes** [comment]
@Ridhwana maybe with the method `validates_each` ? IIRC `ActiveModel::Validations` has been extracted exactly for custom validations. I haven't tried though. If we were to switch to JSONB we could also use JSON schema to validate but that's advanced (see https://dev.to/kputra/rails-postgresql-jsonb-part-1-4ibg#chapter-4)

**Comment by Ridhwana** [comment]
> @Ridhwana maybe with the method `validates_each` ? IIRC `ActiveModel::Validations` has been extracted exactly for custom validations for fields that can't use ActiveRecord validations. I haven't tried though. If we were to switch to JSONB we could also use JSON schema to validate but that's advanced (see https://dev.to/kputra/rails-postgresql-jsonb-part-1-4ibg#chapter-4)

Hmm I see what you mean, let me try it out tomorrow morning and feedback. Thank you @rhymes !

**Comment by citizen428** [comment]
@Ridhwana 

> then we have some config living on the `internal/config` and others on other routes.

I'm not sure that's a problem, `/internal/config` is already getting very long as is and I'd personally have no objections to splitting it up. We're also going to merge `/admin` and `/internal`, so there may be a bigger restructuring in the future anyway.

Regarding `ActiveModel::Validations`: I just made a PR that uses them in case you need inspiration: https://github.com/forem/forem/pull/9534 I'd advise against trying to fit this into `SiteConfig` though. I'd probably add some sort of "service object" for this that takes care of all validations etc. and then persists the whole array when everything is ok. Let me know if you need help with the `ActiveModel` stuff, I've used it quite a bit in the past (e.g. include `ActiveModel::Naming` so form helpers etc. can infer names, etc.).

**Comment by benhalpern** [comment]
@Ridhwana Since the general vibe here is that the comments are "non-blocking", I propose we move forward and get this shipped.

In the vein of low-hanging fruit as we get new forems live, this strikes me as a good thing to get into prod. Let me know if I can be helpful pushing this over the finish line.

**Comment by Ridhwana** [comment]
@benhalpern agreed, I had a chat with @citizen428 during cooldown and we had determined a way forward. I changed priority due to the start of the cycle, but I'll pop on this on my queue next week.  👍🏽

**Comment by mstruve** [comment]
Hey @Ridhwana! What are your thoughts on this PR? Is this something we want to get pushed out or should we close it for now?

**Comment by citizen428** [comment]
@mstruve This is mostly done. It was already mostly done before the last cycle started, but then was "parked" since it's a cooldown project. Given that people inquired about it twice during the previous cycle, we shoul push it over the finish line instead of closing it. I'll see if I'll get to it while @Ridhwana is away next week, if not hopefully she can finish it during the last week of the cooldown.

**Comment by citizen428** [comment]
Looking back at the discussion above, `validates_each` as suggested by @rhymes isn't what we're looking for. According to [the documentation](http://api.rubyonrails.org/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates_each), its purpose is to validate a series of attributes against a block, not multiple objects:

```ruby
Validates each attribute against a block.
```

@Ridhwana I propose you ignore my comment above for now, and we merge a bare-bones first working version of this and then incrementally improve it.

File: app/services/site_configs/update_navigation.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
+module SiteConfigs
+  class UpdateNavigation
+    class NavigationLink
```
Comment (citizen428) [inline app/services/site_configs/update_navigation.rb:3]: Using this class to have a home for the actual validations.

File: app/services/site_configs/update_navigation.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      include ActiveModel::Validations
+
+      URI_REGEXP = URI::DEFAULT_PARSER.make_regexp(%w[http https]).freeze
```
Comment (citizen428) [inline app/services/site_configs/update_navigation.rb:6]: This regex could be better, e.g. `https://` is valid according to it since almost all parts are optional.

File: app/services/site_configs/update_navigation.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      SVG_REGEXP = /\A<svg .*>/i.freeze
+
+      validates :name, :url, :icon, presence: true
+      validates :url, format: URI_REGEXP
+      validates :icon, format: SVG_REGEXP
+
+      def self.from_hash(hash)
```
Comment (citizen428) [inline app/services/site_configs/update_navigation.rb:13]: Custom initializer to simplify creating these objects.

File: app/services/site_configs/update_navigation.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      @navigation_links = navigation_links
+      @errors = {}
+    end
+
+    def call
+      @navigation_links.each_with_index do |attributes, index|
+        navigation_link = NavigationLink.from_hash(attributes)
+        next if navigation_link.valid?
+
+        @errors[index] = navigation_link.errors.to_a
```
Comment (citizen428) [inline app/services/site_configs/update_navigation.rb:42]: The errors are key with their array index, e.g. the errros of the second element will go to index 1.

File: app/views/admin/profile_fields/_add_group_modal.html.erb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-<button class="btn btn-secondary" type="button" onclick="showAddGroupModal()">
+<button class="btn btn-primary" type="button" onclick="showAddGroupModal()">
```
Comment (mstruve) [inline app/views/admin/profile_fields/_add_group_modal.html.erb:1]: This seems a little random, was this change intended?

File: db/migrate/20201001095312_create_navigation_links.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class CreateNavigationLinks < ActiveRecord::Migration[6.0]
+  def change
+    create_table :navigation_links do |t|
```
Comment (mstruve) [inline db/migrate/20201001095312_create_navigation_links.rb:3]:
```
validates :name, :url, :icon, presence: true
```
Can we add these validations at the database level as well?

**Review by mstruve — COMMENTED** [review]
CONGRATS on coming to a simpler solution, I have been down that road before, its always nice to see it all pay off in the end once you have overhauled your strategy. this looks really clean!

File: app/views/admin/profile_fields/_add_group_modal.html.erb (hunk 6 above)
Comment (Ridhwana) [reply app/views/admin/profile_fields/_add_group_modal.html.erb:1]: Yes 🙈, I just slipped it in sneakily because I noticed that all the other add buttons on the admin interface had a primary class name.

File: db/migrate/20201001095312_create_navigation_links.rb (hunk 7 above)
Comment (Ridhwana) [reply db/migrate/20201001095312_create_navigation_links.rb:3]: @mstruve I've updated the migration here to allow for validation and I've just combined the two migration files into one whilst nobody has really pulled down the branch as yet and altered their local db.  Does this look good?

**Comment by Ridhwana** [comment]
> CONGRATS on coming to a simpler solution, I have been down that road before, its always nice to see it all pay off in the end once you have overhauled your strategy. this looks really clean!

**Comment by benhalpern** [comment]
Woohoo, congrats on getting the PR open, taking a deeper look now. I just re-ran the test which seems like an unrelated failure.

**Review by mstruve — APPROVED** [review]
pulled down and tested and this works great!

File: app/models/navigation_link.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class NavigationLink < ApplicationRecord
```
Comment (rhymes) [inline app/models/navigation_link.rb:1]: this is missing one validator: `validates :requires_auth, inclusion: { in: [true, false]}`

Learned recently that is important :D

File: app/controllers/admin/navigation_links_controller.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
+module Admin
+  class NavigationLinksController < Admin::ApplicationController
+    ALLOWED_PARAMS = %i[
+      name url icon requires_auth
+    ].freeze
+    layout "admin"
+
+    def index
+      @navigation_links = NavigationLink.all.order(:name)
```
Comment (rhymes) [inline app/controllers/admin/navigation_links_controller.rb:9]: FYI `.all` is redundant in this case:

```ruby
[5] pry(main)> NavigationLink.all.order(:name).to_sql
=> "SELECT \"navigation_links\".* FROM \"navigation_links\" ORDER BY \"navigation_links\".\"name\" ASC"
[6] pry(main)> NavigationLink.order(:name).to_sql
=> "SELECT \"navigation_links\".* FROM \"navigation_links\" ORDER BY \"navigation_links\".\"name\" ASC"
```

File: app/controllers/admin/navigation_links_controller.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      else
+        flash[:error] = "Error: #{navigation_link.errors_as_sentence}"
+      end
+      redirect_to admin_navigation_links_url
+    end
+
+    private
+
+    def navigation_link_params
+      allowed_params = ALLOWED_PARAMS
```
Comment (rhymes) [inline app/controllers/admin/navigation_links_controller.rb:45]: this alias is probably redudant, you can pass the constant to `.PERMIT`

File: app/views/admin/navigation_links/_add_navigation_link_modal.html.erb [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+                <%= form.label :name %>
+                <%= form.text_field :name, class: "form-control" %>
+              </div>
+              <div class="form-group">
+                <%= form.label :url %>
+                <%= form.text_field :url, class: "form-control" %>
+              </div>
+              <div class="form-group">
+                <%= form.label :icon %>
+                <%= form.text_field :icon, class: "form-control" %>
```
Comment (rhymes) [inline app/views/admin/navigation_links/_add_navigation_link_modal.html.erb:30]: I think we should say that the icon should be a SVG in the explanation, as I obviously used a png and had to start from scratch when presented the error :( 
This should also be a text area I guess...

File: app/models/navigation_link.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+  URI_REGEXP = URI::DEFAULT_PARSER.make_regexp(%w[http https]).freeze
+  SVG_REGEXP = /\A<svg .*>/i.freeze
```
Comment (rhymes) [inline app/models/navigation_link.rb:3]: this is not going to work in all cases, as SVGs can have the XML tags at the start:

```xml
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg>
```

and still be valid. also They are also still valid if they start with a newline:

```xml

<svg ... />
```

It's probably better to just check if the svg tag is in there :D

File: app/models/navigation_link.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+
+  validates :name, :url, :icon, presence: true
+  validates :url, format: URI_REGEXP
```
Comment (rhymes) [inline app/models/navigation_link.rb:6]: why not use the URI validator 😊 ?

For example:

```ruby
validates :url, url: { schemes: %w[https http] }
```

File: app/views/admin/navigation_links/_add_navigation_link_modal.html.erb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+              </div>
+              <div class="form-group">
+                <%= form.check_box :requires_auth %>
+                <%= form.label :requires_auth %>
```
Comment (rhymes) [inline app/views/admin/navigation_links/_add_navigation_link_modal.html.erb:34]: A question: what does requiring authentication mean? Is it authentication or authorization?

File: app/views/admin/navigation_links/_edit_navigation_link_modal.html.erb [hunk 15]
```diff
@@ -0,0 +1,51 @@
+<div onclick="toggleEditNavigationLinkModal('<%= link[:id] %>')" class="color-accent-warning-darker cursor-pointer">
```
Comment (rhymes) [inline app/views/admin/navigation_links/_edit_navigation_link_modal.html.erb:1]: it looks like this is exactly like add except for the title adn the onclick, maybe they can be generalized?

File: spec/models/navigation_link_spec.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+RSpec.describe NavigationLink, type: :model do
+  describe "validations" do
+    describe "presence validations" do
+      it { is_expected.to validate_presence_of(:name) }
+      it { is_expected.to validate_presence_of(:url) }
+      it { is_expected.to validate_presence_of(:icon) }
+    end
+
+    describe "regex vaidations" do
```
Comment (rhymes) [inline spec/models/navigation_link_spec.rb:11]: typo

**Review by rhymes — COMMENTED** [review]
I have a generic question: is there a reason why the creation page is a modal and not a standard page? I tried to add an invalid value and I had to start from scratch when I was presented the error.

Thanks for kickstarting this @Ridhwana. I left a bunch of notes in the review!

File: app/models/navigation_link.rb (hunk 8 above)
Comment (Ridhwana) [reply app/models/navigation_link.rb:1]: ✅

File: app/controllers/admin/navigation_links_controller.rb (hunk 9 above)
Comment (Ridhwana) [reply app/controllers/admin/navigation_links_controller.rb:9]: Urgh I've done this before and you corrected it, and I did it again 🤦🏽‍♀️ ✅

File: app/controllers/admin/navigation_links_controller.rb (hunk 10 above)
Comment (Ridhwana) [reply app/controllers/admin/navigation_links_controller.rb:45]: ✅

File: app/views/admin/navigation_links/_add_navigation_link_modal.html.erb (hunk 11 above)
Comment (Ridhwana) [reply app/views/admin/navigation_links/_add_navigation_link_modal.html.erb:30]: Definitely, in the PR I'm working on currently, I've added an explanation for the page and for the fields. I wanted to get the foundation working in this PR. And thanks, changed to text_area ✅

File: app/views/admin/navigation_links/_add_navigation_link_modal.html.erb (hunk 14 above)
Comment (Ridhwana) [reply app/views/admin/navigation_links/_add_navigation_link_modal.html.erb:34]: its authentication, I'll change this to requires_sign_in, it seems more user friendly to an admin. I'm jsut gonna rollback my migration and update the current one instead of creating a new one as its still in development, you may need to do the same if you've pulled down this branch :( sorry

File: app/models/navigation_link.rb (hunk 13 above)
Comment (Ridhwana) [reply app/models/navigation_link.rb:6]: ✅

File: app/views/admin/navigation_links/_edit_navigation_link_modal.html.erb (hunk 15 above)
Comment (Ridhwana) [reply app/views/admin/navigation_links/_edit_navigation_link_modal.html.erb:1]: I'll do this in a subsequent PR. I've made a note of it.

File: spec/models/navigation_link_spec.rb (hunk 16 above)
Comment (Ridhwana) [reply spec/models/navigation_link_spec.rb:11]: 🙈✅

File: app/models/navigation_link.rb (hunk 12 above)
Comment (Ridhwana) [reply app/models/navigation_link.rb:3]: Changed ✅

Screenshot 2020-10-02 at 13 26 07

**Comment by Ridhwana** [comment]
> I have a generic question: is there a reason why the creation page is a modal and not a standard page? I tried to add an invalid value and I had to start from scratch when I was presented the error.
> 
> Thanks for kickstarting this @Ridhwana. I left a bunch of notes in the review!

Hey @rhymes thanks so much for the review, I updated the PR, can you please review again? 😊 I'm sorry if i messed up your db possibly, I didn't want to add unnecessary migration files so I just rolled it back and created it again. 

The reason I kept it to a modal is because we did something similar for the profile fields and it felt like that was the direction we moving towards. I like the UX of a modal because a user never needs to leave the page to CRUD, but you make a valid point about the errors, its frustrating to lose your input. I wonder if there's some way around that...I'll investigate

**Review by rhymes — APPROVED** [review]
Looks nice! I didn't remember the profile fields page was done with modals, I just expected the page to work like the rest of the admin with the good old pages :P 

Just a note, do you think, a11y wise, that the modals are fine? As @nickytonline has been spending time improving a11y of the front website I think we should keep that in mind when we add features to the backoffice as well

**Comment by Ridhwana** [comment]
> Looks nice! I didn't remember the profile fields page was done with modals, I just expected the page to work like the rest of the admin with the good old pages :P
> 
> Just a note, do you think, a11y wise, that the modals are fine? As @nickytonline has been spending time improving a11y of the front website I think we should keep that in mind when we add features to the backoffice as well

@rhymes this is something I will double check and also speak to Nick about :) Thanks!
